### PR TITLE
[Pallas] Support Flash Attention backward kernels

### PR DESCRIPTION
**Pull request by alanwaketan**
Summary:
This changes refactors custom_kernel.py to support all three new kernels from Pallas that are involved in Flash Attention backward calculations.

The refactoring includes:
1. Adds support for static_argnums which will ignore some positional arguments for jax tracing.
2. Separate jax tracing part out such that we can do the tracing alone.

Test Plan:
PJRT_DEVICE=TPU python test/test_pallas.py

### CODE REVIEW

File: test/test_pallas.py [hunk 1]
```diff
@@ -283,6 +284,158 @@ def add_minus_vectors(x: jax.Array, y: jax.Array) -> jax.Array:
...
+    grad_i = grad_i.unsqueeze(-1).expand(3, 2, 128, MIN_BLOCK_SIZE)
+    grad_q = torch.randn(3, 2, 128, 4).to("xla")
+    torch_xla._XLAC._xla_tpu_custom_call_([grad_q],
+                                          [q, k, v, l, m, grad_o, grad_i],
+                                          payload)
+
+    xm.mark_step()
+
+    # TODO: I don't really know how to test the value. Let's do the shape check for now.
+    self.assertEqual(grad_q.shape, (3, 2, 128, 4))
```
Comment (JackCaoG) [inline test/test_pallas.py:437]: if we do the fwd and do `res.backward` then check the grad on q they should match?

Comment (alanwaketan) [reply test/test_pallas.py:437]: The softmax is done differently. I don't think there is any guarantees.

Comment (JackCaoG) [reply test/test_pallas.py:437]: result should still be somewhat close right? we can tune down the precision. If the result return by this is dramatically different than the one that was computed using dot attention that seems wrong..

Comment (alanwaketan) [reply test/test_pallas.py:437]: https://pytorch.org/docs/stable/generated/torch.nn.Softmax.html

Softmax requires all the elements to produce the results, but flash attention chunks the data into blocks and use a technique called tiling to make sure the softmax still serve the functionality to stable the data. Since there are no aggregation, I don't know how the tiling softmax could produce the same results as the regular one.

In JAX, I have to use `atol=1e-01, rtol=1e-01` to do the comparisons...

**Review by JackCaoG — APPROVED** [review]

**Comment by alanwaketan** [comment]
Thanks, Jack!
